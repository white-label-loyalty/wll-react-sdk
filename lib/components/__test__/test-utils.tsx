import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
