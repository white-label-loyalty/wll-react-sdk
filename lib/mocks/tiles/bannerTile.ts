import {
  BannerTileConfig,
  CTALinkTarget,
  TileHeight,
  TileType,
} from '../../types/tile';

export type BannerTileMockConfig = Partial<BannerTileConfig> & {
  active?: boolean;
  tileHeight?: TileHeight;
};

export const createBannerTileMock = (config?: BannerTileMockConfig) => {
  const {
    active = true,
    tileHeight = TileHeight.Full,
    title = '',
    description = '',
    ctaText = '',
    ctaLink = '',
    ctaLinkTarget = CTALinkTarget.sameWindow,
    artworkUrl = 'https://images.pexels.com/photos/6250883/pexels-photo-6250883.jpeg',
  } = config || {};

  return {
    id: '3',
    type: TileType.Banner,
    active,
    createdAt: '',
    updatedAt: '',
    tileHeight,
    priority: 1,
    configuration: {
      title,
      description,
      ctaText,
      ctaLink,
      ctaLinkTarget,
      artworkUrl,
    },
  };
};
