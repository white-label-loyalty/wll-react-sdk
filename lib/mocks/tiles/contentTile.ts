import {
  ContentTileConfig,
  Tile,
  TileHeight,
  TileType,
} from '../../types/tile';

export type ContentTileMockConfig = Partial<ContentTileConfig> & {
  tileHeight?: TileHeight;
  active?: boolean;
};

const defaultConfig: ContentTileConfig = {
  title: 'Gold Tier Unlocked!',
  body: "You've unlocked exclusive Gold member benefits including 2X points on every purchase, priority support, and VIP event access.",
  artworkUrl:
    'https://images.pexels.com/photos/352097/pexels-photo-352097.jpeg',
};

export const createContentTileMock = (config?: ContentTileMockConfig): Tile => {
  const {
    tileHeight = TileHeight.Full,
    active = true,
    ...tileConfig
  } = config || {};

  const mergedConfig = { ...defaultConfig, ...tileConfig };

  return {
    id: '1',
    type: TileType.Content,
    active,
    createdAt: '',
    updatedAt: '',
    tileHeight,
    priority: 1,
    configuration: mergedConfig,
  };
};
