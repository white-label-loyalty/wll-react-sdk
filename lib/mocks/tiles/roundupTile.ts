import {
  CTALinkTarget,
  RoundupTileConfig,
  TileHeight,
  TileType,
} from '../../types/tile';

export type RoundupTileMockConfig = Partial<RoundupTileConfig> & {
  active?: boolean;
  tileHeight?: TileHeight;
  defaultLocale?: string;
  details?: any[];
  locale?: string;
  balance?: number;
};

export const createRoundupTileMock = (config?: RoundupTileMockConfig) => {
  const {
    active = true,
    tileHeight = TileHeight.Full,
    title = 'Available Points',
    artworkUrl = 'https://example.com/points.png',
    ctaLink,
    ctaLinkTarget,
    defaultLocale = 'en',
    details = [],
    locale = 'en',
    balance,
    amountPrefix,
    amountSuffix,
  } = config || {};

  return {
    id: '1',
    type: TileType.Roundup,
    active,
    createdAt: '2025-01-26T11:13:38Z',
    updatedAt: '2025-01-26T11:13:38Z',
    tileHeight,
    priority: 1,
    configuration: {
      title,
      artworkUrl,
      ctaLink,
      ctaLinkTarget,
      defaultLocale,
      details,
      locale,
      balance,
      amountPrefix,
      amountSuffix,
    },
  };
};
