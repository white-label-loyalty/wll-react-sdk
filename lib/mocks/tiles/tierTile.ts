import {
  ProgressType,
  TierTileConfig,
  TierTileType,
  Tile,
  TileHeight,
  TileType,
} from '../../types/tile';

export type TierTileMockConfig = Partial<TierTileConfig> & {
  tileHeight?: TileHeight;
  priority?: number;
};

const defaultTier = {
  id: '11ff96d2-c87b-4c89-bf5c-996604fc4d28',
  name: 'Bronze',
  description:
    'Get exclusive access to special rewards, and free upgrades when you stay in our hotel.',
  artworkUrl: 'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
  pointsRequirement: 100,
  priority: 0,
  createdAt: '2024-08-06T08:53:24.307Z',
  updatedAt: '2024-08-06T08:53:24.307Z',
};

const defaultConfig: TierTileConfig = {
  progressType: ProgressType.Name,
  locale: 'en',
  type: TierTileType.currentTier,
  emptyDescription: "Oops. You're not in a tier.",
  emptyArtworkUrl: 'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
  pointsToTierPrefix: undefined,
  title: 'Your Tier',
  tierId: '54f0658a-9409-41f7-a508-ca02c6ac23e1',
  tier: defaultTier,
};

export const createTierTileMock = (config?: TierTileMockConfig): Tile => {
  const mergedConfig = {
    ...defaultConfig,
    ...config,
    tier:
      config?.tier === undefined
        ? defaultConfig.tier
        : Object.keys(config.tier).length === 0
          ? {}
          : { ...defaultTier, ...config.tier },
  };

  return {
    id: 'tier-tile-mock',
    type: TileType.Tier,
    active: true,
    createdAt: '2024-08-06T08:53:24.307Z',
    updatedAt: '2024-08-06T08:53:24.307Z',
    tileHeight: config?.tileHeight ?? TileHeight.Full,
    priority: config?.priority ?? 0,
    configuration: mergedConfig,
  };
};
