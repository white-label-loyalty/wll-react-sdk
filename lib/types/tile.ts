export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
