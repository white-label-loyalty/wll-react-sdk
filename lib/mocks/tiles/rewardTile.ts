import { RewardTileConfig, Tile, TileHeight, TileType } from '../../types/tile';

export type RewardTileMockConfig = Partial<RewardTileConfig> & {
  active?: boolean;
  tileHeight?: TileHeight;
};

const defaultConfig = {
  active: true,
  tileHeight: TileHeight.Full,

  rewardId: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
  defaultLocale: 'en-US',
  showArtwork: true,
  showDetails: true,
  showPrice: true,
  id: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
  createdAt: '2023-05-05T15:11:42.342Z',
  updatedAt: '2024-07-17T10:45:04.350Z',
  name: 'Sweet Chilli Lobster Noodles',
  summary: 'Fresh Lobster Noodles with a spicy chilli sauce',
  artworkUrl: 'https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/',
  pictureUrl: 'https://ucarecdn.com/a486d015-b4ab-4831-bfa5-68f5bbbf63b1/',
  value: 800,
  price: 10,
  pointsMultiplier: 1,
  pointsPrefix: '',
  pointsSuffix: '',
  priority: 10,
  purchasable: true,
  type: 'VOUCHER',
  codeType: 'HUMAN',
};

export const createRewardTileMock = (config?: RewardTileMockConfig): Tile => {
  const {
    active = defaultConfig.active,
    tileHeight = defaultConfig.tileHeight,
    ...tileConfig
  } = config || {};

  const mergedConfig = { ...defaultConfig, ...tileConfig };
  const pointsMultiplier = mergedConfig.pointsMultiplier ?? 1;

  // pictureUrl is hydrated to artworkUrl to maintain consistent image handling across all tiles
  const artworkUrl = mergedConfig.artworkUrl;
  const pictureUrl = artworkUrl;

  return {
    id: '39ac25f0-35b3-4668-acba-032dccc63e91',
    type: TileType.Reward,
    active,
    createdAt: '2024-10-17T10:50:12.673Z',
    updatedAt: '2024-10-17T10:50:12.673Z',
    tileHeight,
    priority: 3,
    configuration: {
      rewardId: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
      showArtwork: mergedConfig.showArtwork,
      showDetails: mergedConfig.showDetails,
      showPrice: mergedConfig.showPrice,
      name: mergedConfig.name,
      summary: mergedConfig.summary,
      id: 'f7ce508f-ca52-46ff-bfb7-03e3761feb4a',
      createdAt: '2023-05-05T15:11:42.342Z',
      updatedAt: '2024-07-17T10:45:04.350Z',
      pictureUrl,
      value: 800,
      price: mergedConfig.price,
      pointsPrefix: mergedConfig.pointsPrefix,
      pointsSuffix: mergedConfig.pointsSuffix,
      priority: 10,
      availability: {
        start: '2023-05-05T15:10:31.274Z',
        end: '',
      },
      purchasable: true,
      tier: null,
      category: {
        name: 'Chicken Salad',
        priority: 0,
        type: 'REWARD',
        id: '677b6b88-d97c-4b8d-80f4-88b04c86948f',
        createdAt: '2023-05-05T10:39:57.595Z',
        updatedAt: '2023-05-05T10:39:57.595Z',
        description: null,
        metadata: null,
        pictureUrl:
          'https://ucarecdn.com/9cda5d3c-75f4-48f0-9f3f-1ece144f6136/',
      },
      discounts: [],
      redemptionMessage: null,
      visibilityCriteria: null,
      type: 'VOUCHER',
      codeType: 'HUMAN',
      code: null,
      purchaseExpiration: null,
      hideCode: false,
      notificationConfig: null,
      artworkUrl,
      pointsMultiplier,
    },
  };
};
