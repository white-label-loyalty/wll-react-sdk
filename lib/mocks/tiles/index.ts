import { Tile, TileType } from '../../types/tile';
import { createBadgeTileMock } from './badgeTile';
import { createBannerTileMock } from './bannerTile';
import { createContentTileMock } from './contentTile';
import { createPointsTileMock } from './pointsTile';
import { createRewardCategoryTileMock } from './rewardCategoryTile';
import { createRewardTileMock } from './rewardTile';
export { createContentTileMock } from './contentTile';

export const createTileMock = (type: TileType, config?: any): Tile => {
  switch (type) {
    case TileType.Content:
      return createContentTileMock(config);
    case TileType.Reward:
      return createRewardTileMock(config);
    case TileType.Banner:
      return createBannerTileMock(config);
    case TileType.Points:
      return createPointsTileMock(config);
    case TileType.Badge:
      return createBadgeTileMock(config);
    case TileType.RewardCategory:
      return createRewardCategoryTileMock(config);
    case TileType.Tier:
    // TODO: Implement other tile types as they are needed
    default:
      throw new Error(`Tile type ${type} not yet implemented`);
  }
};
