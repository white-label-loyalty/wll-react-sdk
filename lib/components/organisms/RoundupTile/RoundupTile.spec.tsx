import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import React from 'react';
import { createRoundupTileMock } from '../../../mocks/tiles/roundupTile';
import { TileHeight } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { RoundupTile } from './index';

const defaultTile = createRoundupTileMock();

describe('<RoundupTile />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(<RoundupTile tile={defaultTile} />);
    expect(container).toMatchSnapshot();
  });
});

describe('RoundupTile Rendering States', () => {
  it('returns null when tile is not active', () => {
    const inactiveTile = createRoundupTileMock({ active: false });
    const { container } = render(<RoundupTile tile={inactiveTile} />);
    expect(container).toBeEmptyDOMElement();
  });

  it('renders in full size mode correctly', () => {
    const fullSizeTile = createRoundupTileMock({ tileHeight: TileHeight.Full });
    const { getByTestId } = render(<RoundupTile tile={fullSizeTile} />);
    expect(getByTestId('roundup-tile-media')).toBeInTheDocument();
    expect(getByTestId('roundup-tile-title')).toBeInTheDocument();
    expect(getByTestId('roundup-tile-amount')).toBeInTheDocument();
  });

  it('renders in compact mode correctly', () => {
    const compactTile = createRoundupTileMock({ tileHeight: TileHeight.Half });
    const { getByTestId } = render(<RoundupTile tile={compactTile} />);
    expect(getByTestId('roundup-tile-media')).toBeInTheDocument();
    expect(getByTestId('roundup-tile-title')).toBeInTheDocument();
    expect(getByTestId('roundup-tile-amount')).toBeInTheDocument();
  });
});

describe('<RoundupTile /> Content', () => {
  it('displays the correct title', () => {
    render(<RoundupTile tile={defaultTile} />);
    const titleElement = screen.getByTestId('roundup-tile-title');
    expect(titleElement).toBeInTheDocument();
    expect(titleElement.textContent).toBe('Available Points');
  });

  it('displays the correct zero points value', () => {
    const zeroPointsTile = createRoundupTileMock({ balance: 0 });
    render(<RoundupTile tile={zeroPointsTile} />);
    const valueElement = screen.getByTestId('roundup-tile-value');
    expect(valueElement.textContent).toBe('0');
  });
});
