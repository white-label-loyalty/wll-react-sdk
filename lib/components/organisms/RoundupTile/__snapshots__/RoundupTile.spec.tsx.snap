// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoundupTile /> renders and matches snapshot 1`] = `
<div>
  <article
    aria-label="Available Points"
    class="css-view-175oi2r r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
    role="article"
    style="background-color: rgb(248, 247, 252); flex-direction: column; justify-content: flex-start; align-items: stretch;"
  >
    <div
      aria-label="Roundup tile image for Available Points"
      class="css-view-175oi2r r-alignItems-1awozwy r-backgroundColor-16o1091 r-height-rxcuwo r-justifyContent-1777fci r-marginBottom-6gpygo r-overflow-1udh08x r-width-13qz1uu"
      data-testid="roundup-tile-media"
      role="img"
    >
      <div
        class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-height-1pi2tsx r-position-u8s1d r-width-13qz1uu"
      >
        <div
          class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-4gszlv"
          style="background-image: url(https://example.com/points.png);"
        />
        <img
          alt=""
          class="css-accessibilityImage-9pa8cd"
          draggable="false"
          src="https://example.com/points.png"
        />
      </div>
    </div>
    <div
      class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-paddingInline-3o4zer r-width-13qz1uu"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1h0z5md r-width-13qz1uu"
      >
        <div
          class="css-view-175oi2r r-alignItems-1habvwh r-flexDirection-eqz5dr r-justifyContent-1h0z5md r-flexGrow-16y2uox"
        >
          <div
            aria-label="Available Points"
            class="css-text-146c3p1 r-fontFamily-lhg2x6"
            data-testid="roundup-tile-title"
            dir="auto"
            style="color: rgb(0, 0, 0); font-size: 14px; margin-bottom: 8px;"
          >
            Available Points
          </div>
          <article
            aria-label="Amount: 1000"
            class="css-view-175oi2r"
            data-testid="roundup-tile-amount"
            role="article"
          >
            <div
              class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1h0z5md"
            >
              <div
                class="css-text-146c3p1 r-fontFamily-lhg2x6"
                data-testid="roundup-tile-value"
                dir="auto"
                style="color: rgb(57, 46, 215); font-weight: bold; font-size: 24px;"
              >
                1000
              </div>
            </div>
          </article>
        </div>
        <div
          class="css-view-175oi2r r-justifyContent-1h0z5md r-alignItems-1awozwy r-flexDirection-18u37iz"
        />
      </div>
    </div>
  </article>
</div>
`;
